platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/ConsumerProvidedVariantFinderTest.groovy

```diff
@@ -495,9 +499,108 @@ class ConsumerProvidedVariantFinderTest extends Specification {
         // attributes that are the result of the transform are not compatible with the request
         attributeMatcher.isMatchingCandidate(finalAttributes, requested) >> false
 
+        _ * attributeMatcher.isMatchingCandidate(_, _) >> false
         0 * attributeMatcher._
     }
 
+    @Issue("gradle/gradle#33298")
+    def "long compatible transform chains with no path to source don't lead to combinatorial explosion of explored transform paths"() {
+        def requested = AttributeTestUtil.attributes([usage: "requested"])
+        def unrelatedTransforms = (0..10).collect {
+            registration(
+                AttributeTestUtil.attributes("a$it": "a"),
+                AttributeTestUtil.attributes("a$it": "b"),
+            )
+        }
+        def sourceVariant = variant([usage: "source"])
+
+        given:
+        transformRegistry.registrations >> unrelatedTransforms
+
+        when:
+        def result = transformations.findCandidateTransformationChains([ sourceVariant ], requested)
+
+        then:
+        result.empty
+
+        _ * attributeMatcher.isMatchingCandidate(_, _) >> { c, r -> return fakeIsMatchingCandidate(c, r) }
+    }
+
+    @Issue("gradle/gradle#33298")
+    def "transform can only apply once in a transform chain"() {
+        def by = AttributeTestUtil.attributes([first: "b", second: "y"])
+        def ax = variant([first: "a", second: "x"])
+
+        def a = AttributeTestUtil.attributes(first: "a")
+        def b = AttributeTestUtil.attributes(first: "b")
+        def ay = AttributeTestUtil.attributes([first: "a", second: "y"])
+        def bx = AttributeTestUtil.attributes([first: "b", second: "x"])
+
+        given:
+        transformRegistry.registrations >> [
+            registration(a, b),
+            registration(bx, ay)
+        ]
+
+        when:
+        def result = transformations.findCandidateTransformationChains([ ax ], by)
+
+        then:
+        result.empty
+
+        _ * attributeMatcher.isMatchingCandidate(_, _) >> { c, r -> return fakeIsMatchingCandidate(c, r) }
+    }
+
+    @Issue("gradle/gradle#33298")
+    def "diamond transform graph returns both paths"() {
```

Comment:
I tried a couple of memoization strategies on the `doFindTransformedVariants` algorithm, and this behavior seemed easy to break, while it was not documented in tests